crystal: 1.2 -> 1.7

GeopJr opened the pull request:
###### Description of changes

[Changelog](https://github.com/crystal-lang/crystal/blob/master/CHANGELOG.md?plain=1#L1-L965)

This is a copy of https://github.com/NixOS/nixpkgs/pull/173928 but for `1.2 -> 1.6` with some additional changes.

Changes:

- `libffi` is needed now
- No need to fetch https://github.com/crystal-lang/crystal/pull/11399 as it's merged
- Sets `TMPDIR=$TMP` (used in runtime for `Dir.tempdir`)
- One of the errors https://github.com/NixOS/nixpkgs/pull/173928 faced was due to running the tests (and building the compiler) in release mode [which is (probably) not needed](https://github.com/crystal-lang/crystal/issues/12592#issuecomment-1275389912)
- The other error it faces doesn't seem to occur on 1.5, 1.6
- Disabled `aarch64-linux` as 1.5 is [not yet available](https://dev.alpinelinux.org/archive/crystal/)

Not sure if this should have been a review instead to the original draft but it targets different versions.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

GeopJr committed:
crystal: 1.2 -> 1.6

peterhoeg commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -31,21 +30,21 @@
+    #i686-linux = "linux-i686";
+    #x86_64-darwin = "darwin-x86_64";
+    x86_64-darwin = "darwin-universal";
     aarch64-darwin = "darwin-universal";
```
Instead of leaving things commented out, please put actual comments.

peterhoeg commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -31,21 +30,21 @@
+    x86_64-darwin = "darwin-universal";
     aarch64-darwin = "darwin-universal";
-    aarch64-linux = "linux-aarch64";
+    #aarch64-linux = "linux-aarch64";
```
same here about comments

peterhoeg commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -31,21 +30,21 @@
+    # if arch == archs.aarch64-linux then
+    #   "https://dev.alpinelinux.org/archive/crystal/crystal-${version}-aarch64-alpine-linux-musl.tar.gz"
+    # else
       "https://github.com/crystal-lang/crystal/releases/download/${version}/crystal-${version}-${toString rel}-${arch}.tar.gz";
```
put a real comment here instead

peterhoeg commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -230,48 +216,52 @@ let
+  #  sha256s = {
+  #    aarch64-darwin = "1hrs8cpjxdkcf8mr9qgzilwbg6bakq87sd4yydfsk2f4pqd6g7nf";
+  #  };
+  #};
```
we have 1.2 in NixOS 22.05 so please keep it around. And if was to go, remove the lines instead of just commenting them out.

peterhoeg reviewed (changes requested):
thank you for taking this on. There are a few things related to the version we keep around as well as comments.

peterhoeg commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -230,48 +216,52 @@ let
-    sha256 = "sha256-nyOXhsutVBRdtJlJHe2dALl//BUXD1JeeQPgHU4SwiU=";
-    binary = if isAarch64Darwin then binaryCrystal_1_2 else crystal_1_1;
+  crystal_1_6 = generic {
+    version = "1.6.0";
```
and 1.6.1 is out

peterhoeg commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -163,12 +148,12 @@ let
       LLVM_CONFIG = "${llvmPackages.llvm.dev}/bin/llvm-config";
 
       FLAGS = [
-        "--release"
```
1.6.1 has a fix for this, so it can go

GeopJr replied:
Re-enabled in `makeFlags`.

GeopJr committed:
fix: re-enable 1.0.x - 1.2.x

feat: bump 1.6.x to 1.6.1
fix: fetch 12601 patch for 1.3.x-1.6.0
fix: darwin url for versions < 1.2.0

GeopJr committed:
fix: mark broken if > 1.3.x and 1686-linux

or > 1.5.x and aarch64-linux

GeopJr commented:
I'm not really sure how to deal with the current error.

`aarch64-darwin` is broken for < 1.2.0
`i686-linux` builds are no longer available since 1.3.0
`aarch64-linux` are not yet available for >= 1.5.0

GeopJr commented:
Also I'm still not super-convinced about compiling it in release mode, as [straight-shoota mentioned](https://github.com/crystal-lang/crystal/issues/12592#issuecomment-1275389912):

> And I don't think the compiler itself needs to be built in release mode. (though it might be relevant which LLVM version it uses)

It does take awfully long to compile the compiler AND the compiler spec AND run them when release mode is on.

ento commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -64,12 +67,11 @@ let
       '';
 
-      meta.broken = lib.versionOlder version "1.2.0" && isAarch64Darwin;
+      meta.broken = (lib.versionOlder version "1.2.0" && isAarch64Darwin) || (lib.versionAtLeast version "1.3.0" && stdenv.system == "i686-linux") || (lib.versionAtLeast version "1.5.0" && stdenv.system == "aarch64-linux");
```
> ofborg-eval — This PR does not cleanly list package outputs after merging. 

Would it make sense to set `meta.platforms` of the binary derivations based on the keys of `sha256s`, and let the `generic` derivation inherit `binary.meta.platforms`? If there's no upstream release for a platform, then it seems fine to exclude that from the list of supported platforms, rather than saying that the package is broken.

I made these changes locally:

```diff
diff --git a/pkgs/development/compilers/crystal/default.nix b/pkgs/development/compilers/crystal/default.nix
index 67b9296b..cb3ff680 100644
--- a/pkgs/development/compilers/crystal/default.nix
+++ b/pkgs/development/compilers/crystal/default.nix
@@ -67,7 +67,7 @@ let
         patchShebangs $out/bin/crystal
       '';
 
-      meta.broken = (lib.versionOlder version "1.2.0" && isAarch64Darwin) || (lib.versionAtLeast version "1.3.0" && stdenv.system == "i686-linux") || (lib.versionAtLeast version "1.5.0" && stdenv.system == "aarch64-linux");
+      meta.platforms = lib.attrNames sha256s;
     };
 
   commonBuildInputs = extraBuildInputs: [
@@ -237,13 +237,11 @@ let
       };
 
       meta = with lib; {
-        broken = stdenv.isDarwin;
+        inherit (binary.meta) platforms;
         description = "A compiled language with Ruby like syntax and type inference";
         homepage = "https://crystal-lang.org/";
         license = licenses.asl20;
         maintainers = with maintainers; [ david50407 manveru peterhoeg ];
-        platforms = let archNames = builtins.attrNames archs; in
-          if (lib.versionOlder version "1.2.0") then remove "aarch64-darwin" archNames else archNames;
       };
     })
   );
```

and ran (what I think are) the [steps for reproducing ofborg-eval's test](https://github.com/NixOS/ofborg#running-meta-checks-locally):

```sh
curl -o outpaths.nix https://raw.githubusercontent.com/NixOS/ofborg/released/ofborg/src/outpaths.nix
GC_INITIAL_HEAP_SIZE=4g nix-env -f ./outpaths.nix -qaP --no-name --out-path --arg checkMeta true --show-trace > out-paths
```

Without the changes above, I saw the same error as the one in the [gist that has ofborg-eval's check's details](https://gist.github.com/GrahamcOfBorg/d87d2ff07b17141df6f73f32dfe16f75), "attribute 'aarch64-linux' missing", and with the changes, the command ran long enough that it must've gotten past that part. (I had to terminate it after my machine became near-unresponsive for more than 30 minutes.)

GeopJr replied:
Applied and can confirm it works locally, thanks!

I'll leave this as unresolved, in-case one of the maintainers has any input as I'm not very familiar with nix.

kirillrdy commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -255,23 +267,49 @@ rec {
+  };
+
+  crystal_1_6 = generic {
+    version = "1.6.1";
```
1.6.2 has been released

GeopJr replied:
Done, thanks!

GeopJr committed:
feat: bump 1.6.x to 1.6.2

GeopJr committed:
fix: add another patch for < 1.6.0

so 12601 gets applied successfully

ento committed:
fix: supported platforms per version

peterhoeg commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -163,12 +179,13 @@ let
       LLVM_CONFIG = "${llvmPackages.llvm.dev}/bin/llvm-config";
 
       FLAGS = [
-        "--release"
```
I guess we no longer need to do this as you're patching the versions that do not support it.

Otherwise good to. Once this is reinstated I'll squash and merge.

Thanks a ton @GeopJr !!!

GeopJr replied:
The release flag moved to [`makeFlags`](https://github.com/NixOS/nixpkgs/pull/195606/files#diff-1205fbe1629d11896d8a36ee90fed97d79b6dbdf568bf4201483cdcf62a88318R182) (which has [the same effect as far as I am aware](https://github.com/crystal-lang/crystal/blob/master/Makefile#L33)) from the #173928 PR. Is there another reason to include it in `FLAGS`?

GeopJr commented:
ofborg [still has an issue with the platforms](https://gist.github.com/GrahamcOfBorg/3f8ac98049c4a71d3e30d31206778300) that I am too unfamiliar with Nix to solve

miangraham commented:
Result of `nixpkgs-review pr 195606` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>ameba</li>
    <li>mint</li>
    <li>scry</li>
  </ul>
</details>
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>crystal (crystal_1_6)</li>
    <li>crystal2nix</li>
    <li>crystal_1_0</li>
    <li>crystal_1_1</li>
    <li>crystal_1_2</li>
    <li>crystal_1_5</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>oq</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
  </ul>
</details>

straight-shoota commented on pkgs/development/compilers/crystal/default.nix:
```diff
@@ -98,21 +100,37 @@ let
+      ] ++ lib.optionals (lib.versionAtLeast version "1.3.0" && lib.versionOlder version "1.6.0") [
+        # needed for #12601 to get applied successfully
+        (fetchpatch {
+          url = "https://github.com/crystal-lang/crystal/commit/bf4009dacbf67a63a1cbaffddbdc99549bb70a03.patch";
```
That patch adds FFI support on Windows. What's the relevance of that?

GeopJr replied:
Couldn't apply https://github.com/crystal-lang/crystal/pull/12601 without it on < 1.6.0 (due to the `ffi_spec.cr` changes) - if there's a better alternative I'll gladly change it to that!

straight-shoota replied:
Ah, okay that explains it then.

Not sure if all this backporting of fixes is really worth it, though. FFI bindings are only used for the interpreter, which is still only an experimental feature. That's why these fixes are not backported in upstream.

I'd recommend to just disable the interpreter (and FFI specs) in releases prior to 1.6.2. Even at this point the interpreter is known to have bugs.

GeopJr replied:
Interpreter is not enabled. If only the interpreter depends on ffi, then I guess I can pass `-Dwithout_ffi` instead of deleting the `ffi_spec`, right?

straight-shoota replied:
Yes `-Dwithout_ffi` skips the FFI specs.

straight-shoota replied:
I'm also wondering why specs are built in release mode. That's not really necessary and less efficient.

GeopJr replied:
Both should be done now:
- release mode only for the compiler: https://github.com/NixOS/nixpkgs/pull/195606/commits/5564346d68f6696e18dab41cf3546325e7b4647a
- disable ffi for < 1.6.1: https://github.com/NixOS/nixpkgs/pull/195606/commits/b1be734009b3b9aa217d8561846fce9c9beafb37

miangraham committed:
fix: downstreams

miangraham commented:
On the current ofborg fussing:
```
error: attribute 'aarch64-linux' missing
       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-0/pkgs/development/compilers/crystal/default.nix:61:18:
           61|         sha256 = sha256s.${stdenv.system};
       … while evaluating the attribute 'buildCommand' of the derivation 'crystal-binary-1.5.0'
       … while evaluating the attribute 'nativeBuildInputs' of the derivation 'crystal-1.6.2'
       … while evaluating the attribute 'buildInputs' of the derivation 'kakoune.cr-unstable-2021-11-12'
```

Kakoune-cr is platforms.unix and uses crystal. Crystal 1.6.2 doesn't do aarch64-linux. Can't test here but it looks like another `crystal = crystal_1_2` for kakoune-cr should resolve this one.

GeopJr committed:
fix: make kakoune-cr use crystal 1.2

Co-authored-by: M. Ian Graham <704580+miangraham@users.noreply.github.com>

bcardiff commented:
Thanks @GeopJr for pushing this. I was interested in adding devenv support for Crystal and improving the current state of crystal in nixpkg is very needed for that. 🙇‍♂️ 

I noticed in the opening comment you said

> Disabled aarch64-linux as 1.5 is [not yet available](https://dev.alpinelinux.org/archive/crystal/)

Since 28-Oct 1.5.1 packages are available in that location. If I understand the changes in the PR correctly the comment is outdated but 1.6.x package should be available in aarch64-linux, am I right?

---

On a bigger picture, I wonder what do you think of simplifying how crystal compiler is built to avoid huge chains of crystal_1.0 -> crystal_1.1 -> crystal_1.2 -> ... etc.

Since crystal is bootstrapped it is tempting and it should work that dependency, but things can stop working as the underlying platform is a moving target. The simply rule is that the previous crystal minor version should be able to compile the next compiler. You might be able to skip some, but in general I was thinking of following some recipe for bringing crystal up to date in future releases.

1. Declare the previous crystal binary for all platforms
2. Build crystal compiler from (1) + the patches to make it play nice within Nix

So there are no chains of dependencies. Do you see any downside for this?

GeopJr committed:
fix: disable ffi on < 1.6.1 instead of patching the spec

GeopJr committed:
feat: use release mode only on the compiler and not the specs

GeopJr committed:
feat: 1.5.0 -> 1.5.1 and enable aarch64-linux builds

miangraham commented:
(re: continued ofborg-eval breakage) Hm, it looks like kakoune-cr is still referring to crystal 1.6.2 via shards, even with crystal itself set to 1.2, and breaking on aarch64-linux as a result. Shards gets its version of crystal from nixpkgs [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/shards/default.nix#L3) as normal.

Dropping shards down to crystal 1.2 across the board would likely get past the issue but I'm not sure if that's the best way forward. Could also try and split it per arch.

Anyone experienced with the toolchain have thoughts on how to resolve?

bcardiff commented:
Using crystal_1.2 to build shards should be fine ™️ , specially if that unblocks this. shards hasn't received that many updates and the CI seems to check that can still be built with crystal 1.0.

GeopJr commented:
> Since 28-Oct 1.5.1 packages are available in that location. If I understand the changes in the PR correctly the comment is outdated but 1.6.x package should be available in aarch64-linux, am I right?

Should be re-enabled as of https://github.com/NixOS/nixpkgs/pull/195606/commits/392df0cf61d6d2ea4a370d266a32dad0eb61adda

As for the version chains, I'm way too unfamiliar with Nix to have an opinion on that but from my limited understanding, there's not much we can do to avoid chains if a Crystal package depends on a certain version e.g. Invidious using Kemal < 1.3.0

bcardiff commented:
> As for the version chains

My point is that some crystal compiler packages are using the binary from github or the output of a nix package. For example: 

- [Crystal 1.1 nix package depends on Crystal 1.0 nix package](https://github.com/NixOS/nixpkgs/blob/392df0cf61d6d2ea4a370d266a32dad0eb61adda/pkgs/development/compilers/crystal/default.nix#L283)
- [Crystal 1.6 nix package depends on Crystal 1.5 github binary](https://github.com/NixOS/nixpkgs/blob/392df0cf61d6d2ea4a370d266a32dad0eb61adda/pkgs/development/compilers/crystal/default.nix#L303)

I think the later are better to avoid compiler chains.

straight-shoota commented:
Chain refers to the recursive dependency of each version of the crystal package. In order to build crystal_1_6 you need crystal_1_5 and for that crystal_1_4 and for that crystal_1_3 etc.

That's not really necessary. All releases should be able to bootstrap from crystal_1_0 (every package version could use `binary = binaryCrystal_1_0`). That's not necessarily gonna stay that way for future releases, but there will always be a range of releases that can be built from a previous one.

GeopJr commented:
I see now. Here's my attempt at minimizing the version chain:

```diff
diff --git a/pkgs/development/compilers/crystal/default.nix b/pkgs/development/compilers/crystal/default.nix
index 548c72ca..e0795eb2 100644
--- a/pkgs/development/compilers/crystal/default.nix
+++ b/pkgs/development/compilers/crystal/default.nix
@@ -252,55 +252,17 @@ rec {
     };
   };
 
-  binaryCrystal_1_2 = genericBinary {
-    version = "1.2.0";
-    sha256s = {
-      aarch64-darwin = "1hrs8cpjxdkcf8mr9qgzilwbg6bakq87sd4yydfsk2f4pqd6g7nf";
-    };
-  };
-
-  binaryCrystal_1_5 = genericBinary {
-    version = "1.5.1";
-    sha256s = {
-      x86_64-linux = "sha256-pHXD2Z2+Dy1actRx+iXgPBJLWZ5HM27tdGlztLTXh7w=";
-      #i686-linux = ""; no prebuilt binaries since 1.2.0
-      x86_64-darwin = "sha256-QywvyZIkf2Ztt+VfsVUJRBUQgxpyvro0r/otdrby4JI=";
-      aarch64-linux = "sha256-SdgHd3e5R+/gF7OYz3MEaC7ezI/jcyfpCWmZ3qPeRBI=";
-      aarch64-darwin = "sha256-QywvyZIkf2Ztt+VfsVUJRBUQgxpyvro0r/otdrby4JI=";
-    };
-  };
-
-  crystal_1_0 = generic {
-    version = "1.0.0";
-    sha256 = "sha256-RI+a3w6Rr+uc5jRf7xw0tOenR+q6qii/ewWfID6dbQ8=";
-    binary = binaryCrystal_1_0;
-    extraBuildInputs = [ libatomic_ops ];
-  };
-
-  crystal_1_1 = generic {
-    version = "1.1.1";
-    sha256 = "sha256-hhhT3reia8acZiPsflwfuD638Ll2JiXwMfES1TyGyNQ=";
-    binary = crystal_1_0;
-    extraBuildInputs = [ libatomic_ops ];
-  };
-
   crystal_1_2 = generic {
     version = "1.2.2";
     sha256 = "sha256-nyOXhsutVBRdtJlJHe2dALl//BUXD1JeeQPgHU4SwiU=";
-    binary = if isAarch64Darwin then binaryCrystal_1_2 else crystal_1_1;
+    binary = binaryCrystal_1_0;
     extraBuildInputs = [ libatomic_ops ];
   };
 
-  crystal_1_5 = generic {
-    version = "1.5.1";
-    sha256 = "sha256-twDWnJBLc5tvkg3HvbxXJsCPTMJr9vGvvHvfukMXGyA=";
-    binary = binaryCrystal_1_5;
-  };
-
   crystal_1_6 = generic {
     version = "1.6.2";
     sha256 = "sha256-WgU6Y8ww1IYyB0vd5tXwmWBEL5RiPjHA7YzPd21jlsY=";
-    binary = binaryCrystal_1_5;
+    binary = binaryCrystal_1_0;
   };
 
   crystal = crystal_1_6;
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index a89076f0..3e03b4b9 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -13444,10 +13444,7 @@ with pkgs;
   inherit (callPackages ../development/compilers/crystal {
     llvmPackages = if stdenv.system == "aarch64-darwin" then llvmPackages_11 else llvmPackages_10;
   })
-    crystal_1_0
-    crystal_1_1
     crystal_1_2
-    crystal_1_5
     crystal_1_6
     crystal;

```

They built successfully, with the following warning:
```
In src/uri/encoding.cr:119:25

 119 | String.build { |io| encode(string, io, space_to_plus: space_to_plus) }
                           ^-----
Warning: Deprecated URI.encode:space_to_plus. Use `.encode_path` instead.
```

(I didn't wait for compiler specs however).

From a quick look, nothing uses `crystal_1_0`, `crystal_1_1` & `crystal_1_5` so I got rid of them. I'm unsure about darwin & aarch64 however. There are/were some (undocumented as far as I am aware) conditions just for them:
`if isAarch64Darwin then binaryCrystal_1_2 else crystal_1_1;`
&
```nix
inherit (callPackages ../development/compilers/crystal {
    llvmPackages = if stdenv.system == "aarch64-darwin" then llvmPackages_11 else llvmPackages_10;
  })
```

bcardiff commented:
Some historical facts on crystal versions, llvm and nuances.

The github binaries for crystal 1.2.x were the first ones to be universal binaries for darwin. The can be used to build the compiler in aarch64-darwin without the need of rosetta. I believe was llvm 11 the minimum to build a aarch64-darwin compiler also due to some fixes there.

When crystal 1.2 was released it was tempting to use it to build crystal 1.2 itself and not wait for a 1.3 release in order to use crystal 1.2 to build it. So we used the same version of the released compiler in this case. That's "fine", nothing terrible should happen, but it opens the question of why not do it always. 🤷 

With this facts probably we need at least Crystal 1.2 or newer github binaries to build any crystal version >= 1.3. I'm not sure how crystal 1.0 would work in the context of aarch64-darwin unless rosetta is used.

- LLVM 14 support was added in Crystal 1.4
- LLVM 13 support was added in Crystal 1.2.1
- LLVM 11 is needed for aarch64-darwin as stated before, and it should be fine to use it for simplicity

GeopJr committed:
feat: get rid of unused versions

feat: use 1.2.2 to build 1_2 and 1_6
feat: use llvm 13

GeopJr commented:
Thanks for the info!

So I ended up with using 1.2.2 to build 1.2.2 and 1.6.2 (1.2.1 doesn't have aarch64 linux binaries and 1.2.0 needs llvm < 13) and made llvm 13 the default.

miangraham commented:
Quick followup on ameba: This is in fact broken on current master, which explains why I couldn't build it as-is here, even against crystal 1.2.

```console
$ nix shell --impure github:nixos/nixpkgs#ameba
error: builder for '/nix/store/x54wnih7sr2ljmymy915li6qw1ba2jc7-ameba-1.0.1.drv' failed with exit code 2;
       last 10 log lines:
       > Error target ameba failed to compile:
       > Showing last frame. Use --error-trace for full trace.
       >
       > In src/ameba/inline_comments.cr:98:65
       >
       >  98 | Tokenizer.new(lexer).run { |t| commented = true if t.type.comment? }
       >                                                                 ^-------
       > Error: undefined method 'comment?' for Symbol
```

It's not clear to me how 1.0.1 could have ever worked with nixpkgs' current versions of crystal. [Ameba 1.0 says crystal 1.4 is the minimum](https://github.com/crystal-ameba/ameba/releases/tag/v1.0.0).

Rolling it forward to support crystal 1.6 (as already done) looks like the right answer.

GeopJr commented:
Doesn't look like it ever worked https://github.com/NixOS/nixpkgs/pull/185174/checks?check_run_id=7684715391

GeopJr commented:
By the way, both darwin archs failed here:

x86_64: https://logs.nix.ci/?key=nixos/nixpkgs.195606&attempt_id=0c5c9341-6f46-4abf-acfb-a3b6655cb6f8 (build)

aarch64: https://logs.nix.ci/?key=nixos/nixpkgs.195606&attempt_id=8390320c-bd55-417b-9258-9c32e765256a (specs but also many `was built for newer macOS version (12.0) than being linked (11.0)` warnings)

bcardiff commented:
I was doing a pass on this PR on a Mac M1 OSX 12.5.1.

I did `% nix-build -A crystal .` and got

1. A bunch of `ld: warning: directory not found for option '-L/nix/store/9vrr7g90akidbjjymngsp2rysapyj65c-crystal-1.6.2-lib/crystal'`

2. A compiler spec failure (same as CI). It can be skipped as other specs in the same situation by adding in `postPatch`:

```
        substituteInPlace spec/compiler/loader/unix_spec.cr \
          --replace 'it "parses file paths"' 'pending "parses file paths"'
```

This file exists since Crystal 1.3.0, so it needs to be conditionally applied or it will fails for crystal_1_2. Or prefix the substituteInPlace with `mkdir -p spec/compiler/loader;touch spec/compiler/loader/unix_spec.cr`


I guess we need an upstream patch similar to https://github.com/crystal-lang/crystal/pull/12323 to handle the exact message.

<details>
<summary>output of compiler spec failure</summary>

```
  1) Crystal::Loader .parse parses file paths
     Failure/Error: expect_raises(Crystal::Loader::LoadError, /#{Dir.current}\/foobar\.o.+(No such file or directory|image not found)|Cannot open "#{Dir.current}\/foobar\.o"/) do

       Expected Crystal::Loader::LoadError with message matching /\/private\/tmp\/nix-build-crystal-1.6.2.drv-0\/source\/foobar\.o.+(No such file or directory|image not found)|Cannot open "\/private\/tmp\/nix-build-crystal-1.6.2.drv-0\/source\/foobar\.o"/, got #<Crystal::Loader::LoadError: cannot load /private/tmp/nix-build-crystal-1.6.2.drv-0/source/foobar.o (dlopen(/private/tmp/nix-build-crystal-1.6.2.drv-0/source/foobar.o, 0x0009): tried: '/private/tmp/nix-build-crystal-1.6.2.drv-0/source/foobar.o' (no such file))
       Linker arguments: foobar.o
       Search path: > with backtrace:
         # src/compiler/crystal/loader/unix.cr:84:25 in 'load_file'
         # src/compiler/crystal/loader.cr:39:7 in 'new'
         # src/compiler/crystal/loader/unix.cr:60:7 in 'parse:search_paths'
         # spec/compiler/loader/unix_spec.cr:32:9 in '->'
         # src/spec/example.cr:45:13 in 'internal_run'
         # src/spec/example.cr:33:16 in 'run'
         # src/spec/context.cr:18:23 in 'internal_run'
         # src/spec/context.cr:339:7 in 'run'
         # src/spec/context.cr:18:23 in 'internal_run'
         # src/spec/context.cr:339:7 in 'run'
         # src/spec/context.cr:18:23 in 'internal_run'
         # src/spec/context.cr:156:7 in 'run'
         # src/spec/dsl.cr:220:7 in '->'
         # src/crystal/at_exit_handlers.cr:14:19 in 'run'
         # src/crystal/main.cr:50:14 in 'exit'
         # src/crystal/main.cr:45:5 in 'main'
         # src/crystal/main.cr:127:3 in 'main'

     # spec/compiler/loader/unix_spec.cr:31
```

</details>

Besides that, the compiler seems to work fine in this environment.

Regarding the `ld: warning: object file (...) was built for newer macOS version (12.0) than being linked (11.0)` I believe those will go away if we set `MACOSX_DEPLOYMENT_TARGET = "10.11";` somewhere. This was needed in the shell.nix in crystal repo. I'm surprissed it was not needed so far here, maybe something in the build environment changed along the road and something that was implicit needs to be explicit, not sure.

When I did `% nix-build -A crystal_1_2 .` I got


1. A bunch of `ld: warning: directory not found for option '-L/nix/store/s9qd4hb4mygv40jfvd0waxy7v9a5bc5s-crystal-1.2.2-lib/crystal'`


2. No compiler spec failures

Regarding "Undefined symbols for architecture x86_64: ___darwin_check_fd_set_overflow, referenced from:" error in x86_64 builds the only relevant thing I found is that it might be due to an xcode bug https://openradar.appspot.com/FB7647406 yet is a bug from 2020 Xcode 11.4, I'm not sure how to check if those are the current conditions.

What can I do to help this move forward?

GeopJr committed:
fix: darwin compile errors

Co-authored-by: Brian J. Cardiff <459923+bcardiff@users.noreply.github.com>

GeopJr commented:
I applied the suggested changes however since I have no way of testing darwin, they are mostly done blindly.

> What can I do to help this move forward?

I'll go ahead an invite both you and straight-shoota as collaborators on the origin repo (in-case github hasn't granted you write access as maintainers).

Unrelated, but we should probably also go through the rest of the substitutes to see what's no longer needed, for example I doubt the following is needed considering the linked issue mentions it last occurred on 0.32.1:
```nix
# See https://github.com/crystal-lang/crystal/issues/8629
substituteInPlace spec/std/socket/udp_socket_spec.cr \
  --replace 'it "joins and transmits to multicast groups"' 'pending "joins and transmits to multicast groups"'
```

GeopJr committed:
Merge branch 'master' into update/crystal

will commented:
> I applied the suggested changes however since I have no way of testing darwin, they are mostly done blindly.

I have nix installed on an arm mac right now. I'm not sure how to test things in a nixpkgs pull request though, but if you tell me what to run I can try it out if that helps.

GeopJr committed:
fix: versionOlder => lib.versionOlder

GeopJr commented:
> I have nix installed on an arm mac right now. I'm not sure how to test things in a nixpkgs pull request though, but if you tell me what to run I can try it out if that helps.

I believe you can use `nixpkgs-review pr 195606` (or `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 195606"`). However, ofborg (ci) will test on darwin anyway (it will just take some time).

GeopJr committed:
fix: version limits on unix_spec

bcardiff commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13889,18 +13889,17 @@ with pkgs;
 
   inherit (callPackages ../development/compilers/crystal {
-    llvmPackages = if stdenv.system == "aarch64-darwin" then llvmPackages_11 else llvmPackages_10;
+    llvmPackages = llvmPackages_13;
```
Something I don't get from the x86-darwing error is that it refers to clang-11, but shouldn't it refer to clang-13?

Maybe we could try sticking with llvmPackages_11 here? I'm not sure for the underlying reason, but the llvm package in nixpkgs points to 11.

Or we are missing some path to use llvmPackages_13.clang

GeopJr replied:
:/ Closest discussion I could find: https://discourse.nixos.org/t/nixpkgs-aarch64-darwin-build-m1-uses-wrong-old-clang-how-to-fix/ (in our case x86_64 instead of aarch64)

the suggested change adds clang-unwrapped to PATH https://github.com/NixOS/nixpkgs/blob/6de161729c81dc98e844793cc9c8fda29d5ef62a/pkgs/build-support/libredirect/default.nix#L35-L49

Someone on darwin needs to give it a try otherwise I'll just revert to llvmPackages_11 for x86_64

The nix build definitely downloaded clang-13:
```
...
/nix/store/5b4s502i0ifbivjfb0pwxjlf7vrwgr2v-clang-13.0.1-lib
...
/nix/store/dx1q2d5zh3f45y8amkisy5lbq9b48zsx-clang-13.0.1
...
/nix/store/myqkicxg2zg8wmc84k048hn6fznilqm9-clang-wrapper-13.0.1
...
```
> https://logs.nix.ci/?attempt_id=65e680ba-2ce4-4aae-8375-2b4107302113&key=nixos%2Fnixpkgs.195606

GeopJr committed:
fix(tzdata): prepend to ZONE_SOURCES instead of substituting in place

bcardiff commented:
On macOS 12.6.1 (Monterrey)

```
nixpkgs % uname -a
Darwin galera.local 21.6.0 Darwin Kernel Version 21.6.0: Thu Sep 29 20:12:57 PDT 2022; root:xnu-8020.240.7~1/RELEASE_X86_64 x86_64 i386 MacBookPro16,1 Darwin
```

Building `a979bd9214a622ef444a20e2023d37c268c43ea4` I got

```
% nix-build -A crystal .
these derivations will be built:
  /nix/store/pwlhra7iy35xljqgrb9sqa7i2i4z91m4-crystal-1.6.2.drv
building '/nix/store/pwlhra7iy35xljqgrb9sqa7i2i4z91m4-crystal-1.6.2.drv'...
unpacking sources
unpacking source archive /nix/store/n5zijddfgjn4vrl0nwl7b7h9190wr5ff-source
source root is source
patching sources
applying patch /nix/store/jf31m11h0h5xpmgzpa2a4j0dm7kjsqyh-tzdata.patch
patching file src/crystal/system/unix/time.cr
configuring
no configure script, doing nothing
building
build flags: -j16 SHELL=/nix/store/2198gb5ws3cyma9cxrx3clq6p83781kc-bash-5.1-p16/bin/bash CRYSTAL_CONFIG_VERSION=1.6.2 progress=1 all docs release=1
Using /nix/store/5axyk93lvvxwzjpsgwrihzmzw4lxm6cm-llvm-13.0.1-dev/bin/llvm-config [version= 13.0.1]
clang++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/nix/store/5axyk93lvvxwzjpsgwrihzmzw4lxm6cm-llvm-13.0.1-dev/include -std=c++14 -stdlib=libc++  -fno-exceptions -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
CRYSTAL_CONFIG_BUILD_COMMIT="" CRYSTAL_CONFIG_PATH='$ORIGIN/../share/crystal/src' SOURCE_DATE_EPOCH="1672089671"  CRYSTAL_CONFIG_LIBRARY_PATH='$ORIGIN/../lib/crystal' ./bin/crystal build --single-module -D strict_multi_assign --release --progress --threads 16 -Dwithout_interpreter  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
ld: warning: directory not found for option '-L/nix/store/63z68smi2s24nyzhl9l5iv1qpp3rsgn4-crystal-1.6.2-lib/crystal'
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libiconv.dylib) was built for newer macOS version (10.12) than being linked (10.11)
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libiconv-nocharset.dylib) was built for newer macOS version (10.12) than being linked (10.11)
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libcharset.1.dylib) was built for newer macOS version (10.12) than being linked (10.11)
Undefined symbols for architecture x86_64:
  "___darwin_check_fd_set_overflow", referenced from:
      _select_add in libevent.a(select.o)
      _select_del in libevent.a(select.o)
      _select_dispatch in libevent.a(select.o)
ld: symbol(s) not found for architecture x86_64
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
Error: execution of command failed with code: 1: `clang "${@}" -o /private/tmp/nix-build-crystal-1.6.2.drv-0/tmp.QcGXtyzn7v/private-tmp-nix-build-crystal-1.6.2.drv-0-source-src-ecr-process.cr/macro_run  -rdynamic -L/nix/store/63z68smi2s24nyzhl9l5iv1qpp3rsgn4-crystal-1.6.2-lib/crystal -L/nix/store/6yny71ym1icg45rvfwgn6ip5l4bwksq7-crystal-binary-1.2.2/embedded/lib -lgc -L/nix/store/mykvqcspihpk61a8pw0dfnkm615kmksl-libevent-2.1.12/lib -levent -liconv`
make: *** [Makefile:189: .build/crystal] Error 1
builder for '/nix/store/pwlhra7iy35xljqgrb9sqa7i2i4z91m4-crystal-1.6.2.drv' failed with exit code 2
error: build of '/nix/store/pwlhra7iy35xljqgrb9sqa7i2i4z91m4-crystal-1.6.2.drv' failed
```

We can see that the missing `___darwin_check_fd_set_overflow` is there and that clang-11 is used despite the `llvmPackages = llvmPackages_13` in `all-packages.nix`

On top of this branch I attempted to use llvm 13 clang by setting the `CC` env variable:

```
nixpkgs % git diff
diff --git a/pkgs/development/compilers/crystal/default.nix b/pkgs/development/compilers/crystal/default.nix
index e642e8ee7bc..f8593a2dccf 100644
--- a/pkgs/development/compilers/crystal/default.nix
+++ b/pkgs/development/compilers/crystal/default.nix
@@ -165,6 +165,7 @@ let
         export threads=$NIX_BUILD_CORES
         export CRYSTAL_CACHE_DIR=$TMP
         export MACOSX_DEPLOYMENT_TARGET=10.11
+        export CC=${llvmPackages.clang-unwrapped}/bin/clang
       '';
```

```
nixpkgs % nix-build -A crystal .                      
these derivations will be built:
  /nix/store/d63pfq147brgjdydzp3y0i4x2mf6c89n-crystal-1.6.2.drv
building '/nix/store/d63pfq147brgjdydzp3y0i4x2mf6c89n-crystal-1.6.2.drv'...
unpacking sources
unpacking source archive /nix/store/n5zijddfgjn4vrl0nwl7b7h9190wr5ff-source
source root is source
patching sources
applying patch /nix/store/jf31m11h0h5xpmgzpa2a4j0dm7kjsqyh-tzdata.patch
patching file src/crystal/system/unix/time.cr
configuring
no configure script, doing nothing
building
build flags: -j16 SHELL=/nix/store/2198gb5ws3cyma9cxrx3clq6p83781kc-bash-5.1-p16/bin/bash CRYSTAL_CONFIG_VERSION=1.6.2 progress=1 all docs release=1
Using /nix/store/5axyk93lvvxwzjpsgwrihzmzw4lxm6cm-llvm-13.0.1-dev/bin/llvm-config [version= 13.0.1]
clang++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/nix/store/5axyk93lvvxwzjpsgwrihzmzw4lxm6cm-llvm-13.0.1-dev/include -std=c++14 -stdlib=libc++  -fno-exceptions -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
CRYSTAL_CONFIG_BUILD_COMMIT="" CRYSTAL_CONFIG_PATH='$ORIGIN/../share/crystal/src' SOURCE_DATE_EPOCH="1672064457"  CRYSTAL_CONFIG_LIBRARY_PATH='$ORIGIN/../lib/crystal' ./bin/crystal build --single-module -D strict_multi_assign --release --progress --threads 16 -Dwithout_interpreter  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
ld: warning: directory not found for option '-L/nix/store/6ci2jz3hm2q498x5yq2zfd4jih1jhr54-crystal-1.6.2-lib/crystal'
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libiconv.dylib) was built for newer macOS version (10.12) than being linked (10.11)
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libiconv-nocharset.dylib) was built for newer macOS version (10.12) than being linked (10.11)
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libcharset.1.dylib) was built for newer macOS version (10.12) than being linked (10.11)
Undefined symbols for architecture x86_64:
  "___darwin_check_fd_set_overflow", referenced from:
      _select_add in libevent.a(select.o)
      _select_del in libevent.a(select.o)
      _select_dispatch in libevent.a(select.o)
ld: symbol(s) not found for architecture x86_64
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
Error: execution of command failed with code: 1: `/nix/store/dx1q2d5zh3f45y8amkisy5lbq9b48zsx-clang-13.0.1/bin/clang "${@}" -o /private/tmp/nix-build-crystal-1.6.2.drv-0/tmp.rOLzPfAfOq/private-tmp-nix-build-crystal-1.6.2.drv-0-source-src-ecr-process.cr/macro_run  -rdynamic -L/nix/store/6ci2jz3hm2q498x5yq2zfd4jih1jhr54-crystal-1.6.2-lib/crystal -L/nix/store/6yny71ym1icg45rvfwgn6ip5l4bwksq7-crystal-binary-1.2.2/embedded/lib -lgc -L/nix/store/mykvqcspihpk61a8pw0dfnkm615kmksl-libevent-2.1.12/lib -levent -liconv`
make: *** [Makefile:189: .build/crystal] Error 1
builder for '/nix/store/d63pfq147brgjdydzp3y0i4x2mf6c89n-crystal-1.6.2.drv' failed with exit code 2
error: build of '/nix/store/d63pfq147brgjdydzp3y0i4x2mf6c89n-crystal-1.6.2.drv' failed
```

We can see that `clang-13`  was used instead of `clang-11`, yet the missing `___darwin_check_fd_set_overflow` is still there.

If I downgrade `llvmPackages` to `llvmPackages_10` as follows (to mimic how it was before this PR for x86_64-darwin),

```
% git diff
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 3b22223300c..22b4a8ae828 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -13889,7 +13889,7 @@ with pkgs;
   };
 
   inherit (callPackages ../development/compilers/crystal {
-    llvmPackages = llvmPackages_13;
+    llvmPackages = llvmPackages_10;
   })
```

I still get the `___darwin_check_fd_set_overflow` error :-(

```
% nix-build -A crystal .
these derivations will be built:
  /nix/store/dafwli8r6akrf6n8hlprazmnlwl9k5a7-crystal-1.6.2.drv
building '/nix/store/dafwli8r6akrf6n8hlprazmnlwl9k5a7-crystal-1.6.2.drv'...
unpacking sources
unpacking source archive /nix/store/n5zijddfgjn4vrl0nwl7b7h9190wr5ff-source
source root is source
patching sources
applying patch /nix/store/jf31m11h0h5xpmgzpa2a4j0dm7kjsqyh-tzdata.patch
patching file src/crystal/system/unix/time.cr
configuring
no configure script, doing nothing
building
build flags: -j16 SHELL=/nix/store/2198gb5ws3cyma9cxrx3clq6p83781kc-bash-5.1-p16/bin/bash CRYSTAL_CONFIG_VERSION=1.6.2 progress=1 all docs release=1
Using /nix/store/081pffb2aa5ksmav3302cx25l160kyiy-llvm-10.0.1-dev/bin/llvm-config [version= 10.0.1]
clang++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/nix/store/081pffb2aa5ksmav3302cx25l160kyiy-llvm-10.0.1-dev/include -std=c++14 -stdlib=libc++  -fno-exceptions -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
CRYSTAL_CONFIG_BUILD_COMMIT="" CRYSTAL_CONFIG_PATH='$ORIGIN/../share/crystal/src' SOURCE_DATE_EPOCH="1672093091"  CRYSTAL_CONFIG_LIBRARY_PATH='$ORIGIN/../lib/crystal' ./bin/crystal build --single-module -D strict_multi_assign --release --progress --threads 16 -Dwithout_interpreter  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
ld: warning: directory not found for option '-L/nix/store/9va99kkpm2k5lryf10p1rsbi699q715r-crystal-1.6.2-lib/crystal'
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libiconv.dylib) was built for newer macOS version (10.12) than being linked (10.11)
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libiconv-nocharset.dylib) was built for newer macOS version (10.12) than being linked (10.11)
ld: warning: dylib (/nix/store/z1d2x7ybr06vi5czzr27vsd5j77sc6qw-libiconv-50/lib/libcharset.1.dylib) was built for newer macOS version (10.12) than being linked (10.11)
Undefined symbols for architecture x86_64:
  "___darwin_check_fd_set_overflow", referenced from:
      _select_add in libevent.a(select.o)
      _select_del in libevent.a(select.o)
      _select_dispatch in libevent.a(select.o)
ld: symbol(s) not found for architecture x86_64
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
Error: execution of command failed with code: 1: `clang "${@}" -o /private/tmp/nix-build-crystal-1.6.2.drv-0/tmp.N6TcXL8fwc/private-tmp-nix-build-crystal-1.6.2.drv-0-source-src-ecr-process.cr/macro_run  -rdynamic -L/nix/store/9va99kkpm2k5lryf10p1rsbi699q715r-crystal-1.6.2-lib/crystal -L/nix/store/6yny71ym1icg45rvfwgn6ip5l4bwksq7-crystal-binary-1.2.2/embedded/lib -lgc -L/nix/store/mykvqcspihpk61a8pw0dfnkm615kmksl-libevent-2.1.12/lib -levent -liconv`
make: *** [Makefile:189: .build/crystal] Error 1
builder for '/nix/store/dafwli8r6akrf6n8hlprazmnlwl9k5a7-crystal-1.6.2.drv' failed with exit code 2
error: build of '/nix/store/dafwli8r6akrf6n8hlprazmnlwl9k5a7-crystal-1.6.2.drv' failed
```

I think the problem might be on libevent package actually. The lib is already built and references `___darwin_check_fd_set_overflow` in x86-64_darwin . 

Given that https://github.com/curl/curl/issues/5210 is similar and it was resolved eventually with an xcode upgrade maybe the issue here is how libevent was built on x86-64_darwin initially.

I attempted to change something in the libevent formula to trigger a rebuilt locally

```
% git diff
diff --git a/pkgs/development/libraries/libevent/default.nix b/pkgs/development/libraries/libevent/default.nix
index bd5edec68a0..9f2cc1a2fea 100644
--- a/pkgs/development/libraries/libevent/default.nix
+++ b/pkgs/development/libraries/libevent/default.nix
@@ -20,8 +20,8 @@ stdenv.mkDerivation rec {
     })
   ];
 
-  preConfigure = lib.optionalString (lib.versionAtLeast stdenv.hostPlatform.darwinMinVersion "11") ''
-    MACOSX_DEPLOYMENT_TARGET=10.16
+  preConfigure = ''
+    MACOSX_DEPLOYMENT_TARGET=10.11
   '';
```

Unfortunately, the `___darwin_check_fd_set_overflow` error is still present after that.

I am not sure how to maybe tweak which xcode version should be used.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/unable-to-build-crystal-on-x86-64-darwin/24409/1

GeopJr committed:
feat: 1.7.0

GeopJr committed:
Merge branch 'master' of https://github.com/NixOS/nixpkgs into update/crystal

GeopJr committed:
feat: 1.7.1

GeopJr committed:
fix(darwin): use sdk 11

GeopJr commented:
(At the time of writing this)

While I made the Crystal package to use sdk 11 in an attempt to get rid of `was built for newer macOS version (10.12) than being linked (10.11)`, it seems to have actually fixed the x86_64 darwin issue and is currently running the compiler specs for it!

Unfortunately the `was built for newer macOS version (10.12) than being linked (10.11)` did not get fixed.

GeopJr commented:
It finished! So our only issue now is `was built for newer macOS version (10.12) than being linked (10.11)` (though, not critical)

bcardiff reviewed (approved):
Awesome! @GeopJr it's great you found about stdenv override. How did you manage?

I think is not critical and it would be great if we have a non-broken crystal in nixpkgs 😊

GeopJr commented:
> How did you manage?

Quite the rabbit hole - I was trying to find more info on `was built for newer macOS version (10.12) than being linked (10.11)` on this repo, searching `10.12` brought me to this issue: https://github.com/NixOS/nixpkgs/issues/101229 and after reading the comments and linked PRs I saw that zig and go used it so I gave it try (didn't expect it to fix the x86_64 darwin issue)!

> I think is not critical and it would be great if we have a non-broken crystal in nixpkgs

Awesome! I'll run a review to see if anything is broken on 1.7.1 and then try to move this forward!

miangraham commented:
Result of `nixpkgs-review pr 195606` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>ameba</li>
    <li>crystal2nix</li>
    <li>mint</li>
    <li>oq</li>
    <li>thicket</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>crystal (crystal_1_7)</li>
    <li>crystal_1_2</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
  </ul>
</details>

GeopJr commented:
Thanks for the review @miangraham!

Unless someone on darwin wants to debug the failing packages, I'm ok with leaving them as is (if allowed), considering they haven't been built in darwin for ages and might need some workarounds from their maintainers.

This doesn't seem to be a Crystal version issue: ameba, oq, thicket, crystal2nix should work on 1.7 and mint is locked to 1.2.
